org.osgi.util.converter/src/org/osgi/util/converter/ConvertingImpl.java

```diff
@@ -332,6 +344,42 @@ private <T> T convertToCollectionType(InternalConverter converter) {
 		return convertToCollection(converter);
 	}
 
+	@SuppressWarnings("unchecked")
+	private <T> T convertToOptionalType(InternalConverter converter) {
+
+		if (typeArguments != null) {
+			Object o = converter.convert(object).to(typeArguments[0]);
+			return (T) Optional.ofNullable(o);
+		}
+
+		return (T) Optional.ofNullable(object);
+
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> T convertToOptionalInt(InternalConverter converter) {
+
+		int i = converter.convert(object).to(int.class);
+		return (T) OptionalInt.of(i);
+
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> T convertToOptionalDouble(InternalConverter converter) {
+
+		double d = converter.convert(object).to(double.class);
+		return (T) OptionalDouble.of(d);
+
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> T convertToOptionalLong(InternalConverter converter) {
+
+		long l = converter.convert(object).to(long.class);
```

Comment:
The same issue as above. We should use a `Long` in conversion, do a null check, and use `longValue()` when wrapping.